### PR TITLE
:sparkles: Use ErrorIfCRDPathMissing in EnvTest

### PR DESCRIPTION
**Pull request by jamielennox**
The error message if the path to CRDs is missing is misleading.
envtest.Environment has an option to throw a sensible error here if the
path is wrong and for kubebuilder it would be likely we always want to
upload our CRDs.

### CODE REVIEW

File: pkg/plugin/v2/scaffolds/internal/templates/controller/controller_suitetest.go
Comment (prafull01) [inline pkg/plugin/v2/scaffolds/internal/templates/controller/controller_suitetest.go:152]: Hi @jamielennox Thanks for your contribution. :+1: 
Couple of changes:
1. This change should only happen for v3. We have decided to include only bug fixes for v2 going forward. Could you please revert the change done in v2.
2. You need to run `make generate` command locally, which generates the testdata (based on your change) and push those changes as well.

**Review by prafull01 — CHANGES_REQUESTED** [review]

File: pkg/plugin/v3/scaffolds/internal/templates/config/controller/controller_suitetest.go
Comment (Adirio) [inline pkg/plugin/v3/scaffolds/internal/templates/config/controller/controller_suitetest.go:155]:
```suggestion
		CRDDirectoryPaths:     []string{filepath.Join({{ .CRDDirectoryRelativePath }}, "config", "crd", "bases")},
```

**Review by Adirio — CHANGES_REQUESTED** [review]

File: pkg/plugin/v3/scaffolds/internal/templates/config/controller/controller_suitetest.go
Comment (jamielennox) [reply pkg/plugin/v3/scaffolds/internal/templates/config/controller/controller_suitetest.go]: Yep, caught that when i saw the generated code. Thanks.

File: pkg/plugin/v2/scaffolds/internal/templates/controller/controller_suitetest.go
Comment (jamielennox) [reply pkg/plugin/v2/scaffolds/internal/templates/controller/controller_suitetest.go]: Done, thanks.

**Comment by jamielennox** [comment]
I'm not sure how you handle changes here. You can squash the PR or i can merge them and force push. Let me know.

**Comment by prafull01** [comment]
Hi @jamielennox You need to squash the commits into 1 commit and force push. Thanks

**Review by Adirio — APPROVED** [review]
Code-wise, the change LGTM. I don't know what is implied by this flag so someone with better knowledge about `EnvTest` should aprove it.

/lgtm

**Comment by prafull01** [comment]
It is a good to have functionality, when the CRD path specified envtest is not present, it throws error (which it should be). With current implementation even if the CRD is not present, the scaffolded suite doesn't throw error.
I tested it locally, `kb init` and `kb create api`, these two command will scaffold the suite_test.go but does not generate the `config/crd/bases` folder until I run `make test all`:

Running tests before `make manifests` without these changes:
```
prafull@EMPID18004:~/go/src/github.com/prafull01/kubebuilder/proj$ go test -v ./controllers/...
=== RUN   TestAPIs
Running Suite: Controller Suite
===============================
Random Seed: 1604307472
Will run 0 of 0 specs



Ran 0 of 0 Specs in 4.208 seconds
SUCCESS! -- 0 Passed | 0 Failed | 0 Pending | 0 Skipped
--- PASS: TestAPIs (4.21s)
PASS
ok  	sigs.k8s.io/kubebuilder/controllers	4.218s
```

Running tests before `make manifests` with these changes:
```
prafull@EMPID18004:~/go/src/github.com/prafull01/kubebuilder/proj$ go test -v ./controllers/...
=== RUN   TestAPIs
Running Suite: Controller Suite
===============================
Random Seed: 1604307507
Will run 0 of 0 specs

STEP: bootstrapping test environment
2020-11-02T14:28:27.418+0530	DEBUG	controller-runtime.test-env	starting control plane	{"api server flags": []}
2020-11-02T14:28:32.630+0530	DEBUG	controller-runtime.test-env	installing CRDs
Failure [5.212 seconds]
[BeforeSuite] BeforeSuite 
/home/prafull/go/src/github.com/prafull01/kubebuilder/proj/controllers/suite_test.go:52

  Unexpected error:
      <*os.PathError | 0xc0006711a0>: {
          Op: "stat",
          Path: "../config/crd/bases",
          Err: 0x2,
      }
      stat ../config/crd/bases: no such file or directory
  occurred

  /home/prafull/go/src/github.com/prafull01/kubebuilder/proj/controllers/suite_test.go:63
------------------------------


Ran 0 of 0 Specs in 5.246 seconds
FAIL! -- 0 Passed | 0 Failed | 0 Pending | 0 Skipped
```

Running test with these changes after `make manifests` which will generate the CRD:
```
prafull@EMPID18004:~/go/src/github.com/prafull01/kubebuilder/proj$ go test -v ./controllers/...
=== RUN   TestAPIs
Running Suite: Controller Suite
===============================
Random Seed: 1604307565
Will run 0 of 0 specs



Ran 0 of 0 Specs in 7.440 seconds
SUCCESS! -- 0 Passed | 0 Failed | 0 Pending | 0 Skipped
--- PASS: TestAPIs (7.44s)
PASS
ok  	sigs.k8s.io/kubebuilder/controllers	7.453s
```
IMO, this change is LGTM. 

PTAL for
/cc @camilamacedo86

File: pkg/plugin/v3/scaffolds/internal/templates/config/controller/controller_suitetest.go
Comment (camilamacedo86) [inline pkg/plugin/v3/scaffolds/internal/templates/config/controller/controller_suitetest.go:156]: What happens if we create a controller without CRD? 

> $ kubebuilder create api --group ship --version v1beta1 --kind Frigate
> Create Resource [y/n]
> n
> Create Controller [y/n]
> y

Comment (Adirio) [reply pkg/plugin/v3/scaffolds/internal/templates/config/controller/controller_suitetest.go:156]: Just to explain a bit more what Camila is asking, controllers without a CRD can be used for builtin resources for example, or for resources that may be defined by other frameworks. So, technically, we could have a controller without its resource. So imagine we want to create a controller for `v1.Namespace` with no CRD (as the resource is already defined by k8s). Will this flag make the suite fail? In that case we need to set it only in the case we are not deploying a controller for an existing resource.

Comment (camilamacedo86) [reply pkg/plugin/v3/scaffolds/internal/templates/config/controller/controller_suitetest.go:156]: Also see: https://github.com/kubernetes-sigs/kubebuilder/issues/1767
IMO we need to provide the fix first to be able to check this one.

**Comment by camilamacedo86** [comment]
/hold 

Until the fix https://github.com/kubernetes-sigs/kubebuilder/pull/1770 get merged and this one realised with master.

**Comment by camilamacedo86** [comment]
HI @jamielennox, 

Could you please rebase it with the master, run make generate and push? 

/hold
cancel

**Comment by jamielennox** [comment]
I've very simply rebased this, reading #1770 maybe we want to only set true on WireResource?

**Comment by Adirio** [comment]
> I've very simply rebased this, reading #1770 maybe we want to only set true on WireResource?

Probably yes

**Comment by camilamacedo86** [comment]
Hi @jamielennox, 

Could you check what will be the behaviour faced after this change in the scenario https://github.com/kubernetes-sigs/kubebuilder/pull/1765#discussion_r515853290? Could you please let us know?

**Comment by camilamacedo86** [comment]
Hi @jamielennox, 

Could you please get it rebased with the master for we are able to move forward with it?

File: pkg/plugin/v3/scaffolds/internal/templates/controllers/controller_suitetest.go
Comment (camilamacedo86) [inline pkg/plugin/v3/scaffolds/internal/templates/controllers/controller_suitetest.go:163]: We need to do the same for webhooks_suitetest as well.

Comment (jamielennox) [reply pkg/plugin/v3/scaffolds/internal/templates/controllers/controller_suitetest.go:163]: This is more difficult on webhook_suitetest. There's no doResource/WireResource boolean that can be used, the closest point i can see where a doResource is actually used is harcoded to false: https://github.com/kubernetes-sigs/kubebuilder/blob/21de76043e6e14fd775729452c3c2a1a3432ad8e/pkg/plugins/golang/v3/webhook.go#L131 so i can just hardcode it to false as well or ignore it

File: pkg/plugins/golang/v3/scaffolds/internal/templates/api/webhook_suitetest.go
Comment (camilamacedo86) [inline pkg/plugins/golang/v3/scaffolds/internal/templates/api/webhook_suitetest.go:167]: Could we create a variable to pass here and add on top of this var in the sea above the following comment?

// todo: currently is not possible to know if an API was or not scaffolded. We can fix it when  https://github.com/kubernetes-sigs/kubebuilder/issues/1826 be addressed

Comment (jamielennox) [reply pkg/plugins/golang/v3/scaffolds/internal/templates/api/webhook_suitetest.go]: done, i'm sure WireResource is a bad name in webhook, but it's closest to standard i can see.

**Review by camilamacedo86 — APPROVED** [review]
/approve

**Review by camilamacedo86 — APPROVED** [review]
It receives approval of 2 very active contributors/maintainers (@prafull01 and @Adirio) 
The changes here shows fine for me as well. 

Great job @jamielennox 🥇

**Pull request merged**
